apps/ae_backend_service/lib/backend_session.ex

```diff
@@ -67,16 +73,109 @@ defmodule BackendSession do
+    {responder_pub, _priv} = keypair_responder()
+    {_role, _channel_config, _reestablish, initiator_keypair} = state.params
+    {initiator_pub, _priv} = initiator_keypair.()
+
+    fun =
+      &SocketConnector.new_contract(
+        &1,
+        state.responder_contract,
+        [
+          to_charlist(responder_pub),
+          to_charlist(initiator_pub),
+          '15'
+        ],
+        10
+      )
+
+    SessionHolder.run_action(state.pid_session_holder, fun)
+    {:noreply, state}
+  end
+
+  def handle_cast({:channels_update, 2, round_initiator, "channels.update"} = _message, state)
+      when round_initiator in [:self] do
+    {responder_pub, _priv} = keypair_responder()
+    {_role, _channel_config, _reestablish, initiator_keypair} = state.params
+    {initiator_pub, _priv} = initiator_keypair.()
+
+    some_salt = ContractHelper.add_quotes("some_salt")
+    coin = ContractHelper.add_quotes("heads")
+
+    fun =
+      &SocketConnector.call_contract_dry(
+        &1,
+        state.responder_contract,
+        'compute_hash',
+        [to_charlist(some_salt), to_charlist(coin)],
+        &2
+      )
+
+    {:ok, {:bytes, [], hash}} = SessionHolder.run_action_sync(state.pid_session_holder, fun)
+    Logger.error("hash is #{inspect(hash)}")
+
+    fun1 =
+      &SocketConnector.call_contract(
+        &1,
+        state.responder_contract,
+        'provide_hash',
+        [to_charlist(ContractHelper.to_sophia_bytes(hash))]
+      )
+
+    SessionHolder.run_action(state.pid_session_holder, fun1)
+    {:noreply, %__MODULE__{state | game: %{hash: hash, coin: coin, salt: some_salt}}}
+  end
+
+  def handle_cast({:channels_update, 3, round_initiator, "channels.update"} = _message, state)
+      when round_initiator in [:self] do
+    Logger.info("Waiting for player to make a guess")
+    {:noreply, state}
+  end
+
+  def handle_cast(
+        {:channels_update, 4, round_initiator, "channels.update"} = _message,
+        %__MODULE__{game: game} = state
+      )
+      when round_initiator in [:other] do
+    Logger.info("Other player made a guess")
+
+    fun =
+      &SocketConnector.call_contract(
+        &1,
+        state.responder_contract,
+        'reveal',
+        [to_charlist(game.salt), to_charlist(game.coin)]
+      )
+
+    SessionHolder.run_action(state.pid_session_holder, fun)
+    {:noreply, state}
+  end
+
+  def handle_cast({:channels_update, 5, round_initiator, "channels.update"} = _message, state)
+      when round_initiator in [:self] do
+    Logger.info("Game end, backend emptying contract")
+
+    fun =
+      &SocketConnector.call_contract(
+        &1,
+        state.responder_contract,
+        'drain',
+        []
+      )
+
+    SessionHolder.run_action(state.pid_session_holder, fun)
+
+    {:noreply, state}
+  end
 
-    fun = &SocketConnector.new_contract(&1, responder_contract, 10)
+  def handle_cast({:channels_update, 6, _round_initiator, "channels.update"} = _message, state) do
```

Comment:
You are right. This will change today. Also a new contract will be added. We need to think about branching strategies, it's not evident that the any "master" user is interested in starting a toin-coss game. At least docs should be updated.